Add flag to print unfolded projections as cases (#16994)
coq/coq: doc/sphinx/language/core/records.rst

.. flag:: Printing Primitive Projection Parameters

>    This compatibility :term:`flag` (off by default) reconstructs
>    internally omitted parameters at printing time (even though they
>    are absent in the actual AST manipulated by the kernel).

Reviewer: ```suggestion
   When set, internally omitted parameters are reconstructed at printing time
   even though they are absent in the AST manipulated by the kernel.  Default is unset.
```

Author: Not interested in touching this more in this PR

allows to precisely mimic the usual unfolding rules of :term:`constants <constant>`.
Projections obey the usual ``simpl`` flags of the :cmd:`Arguments`
command in particular.

Unfolded primitive projections can be built using the compatibility
match syntax for primitive records.

> .. flag:: Printing Unfolded Projection As Projection

Reviewer 2: How about `Printing Wrap Primitive Projections` or `Printing Use Constants for Primitive Projections`? "As Projection" made me think of how the opposite to what the behavior seems to be.

Author: What do constants have to do with it? Not sure what you mean by wrapping either.
btw if there's a better name for the negation of the behaviour we can make that the flag

Reviewer 2: Hmm, maybe I'm misunderstanding what this change is about. Is there yet another way of how primitive projections can occur beyond the two listed in https://coq.inria.fr/refman/language/core/records.html#compatibility-projections-and-match (constants and unfolded primitive projections)? The terminology in that documentation paragraph makes sense to me; I was trying to suggest something in the same style from memory.

Reviewer 2: How about `Printing Primitive Projections As Matches` or `Printing Primitive Projections As Compatibility Constants`?

Author: There is also the folded primitive projection
Typically
~~~
Set Primitive Projections.
Record R param := r { p : T }.

Check Eval cbv zeta in let f := @p in f _ v. (* @p param v, this is the constant *)
Check @p _ v. (* @p _ v, this is the folded primitive projection *)
Print p. (* = fun param x => @p _ x, the @p is the unfolded primitive projection *)
~~~
I don't remember how to tell constant and primitive projection apart when there are no parameters but I think there was a way with some printing option.

Reviewer 2: :scream: I now understand the option name you proposed, thanks. `Printing Unfolded Projection As Folded Projection` or `Printing Unfolded Projection As Match` might be better, but either way probably a bigger improvement would come from explaining what's up with these three ways in the manual. Currently, [searching for unfolded primitive projection](https://coq.inria.fr/refman/search.html?q=unfolded+primitive+projection&check_keywords=yes&area=default) returns no full matches in the refman.

Author: >Currently, searching for unfolded primitive projection returns no full matches in the refman.

There's a search issue, it appears eg in https://coq.inria.fr/refman/language/core/records.html#compatibility-projections-and-match

I don't know how sphinx search works so I don't know if there's some invocation that will make it do the desired thing.

Reviewer 3: See also #12811.

allows to precisely mimic the usual unfolding rules of :term:`constants <constant>`.
Projections obey the usual ``simpl`` flags of the :cmd:`Arguments`
command in particular.

Unfolded primitive projections can be built using the compatibility
match syntax for primitive records, or by reducing the compatibility constant.

User-written :g:`match` constructs on primitive records are
desugared using the unfolded primitive projections and `let` bindings.

.. flag:: Printing Unfolded Projection As Match

>    By default this flag is off and unfolded primitive projections are
>    printed the same as folded primitive projections. By unsetting this
>    flag, unfolded primitive projections are instead printed as
>    let-style matches ``let '{| p := p |} := c in p``.

Reviewer: ```suggestion
   By default this flag is off and unfolded primitive projections are
   printed the same as folded primitive projections. By setting this
   flag, unfolded primitive projections are instead printed as
   let-style matches in the form ``let '{| p := p |} := c in p``.
```
